### DISCUSSION TITLE
How can you mock the internal fetch of your route handler?

### DISCUSSION THREAD

**Discussion by lucasmotta** [Xunnamius/next-test-api-route-handler#1063] [Q&A] [answered]
Hey there, thanks for such an amazing library. I'm having some issues trying to mock the response of a fetch used inside the route handler.

Let's say I've a route handler like this (simplified for demo purposes):

```tsx
// [token]/route.js
export const GET = (request, { params }) => {
  try {
    if (!params.token) {
      throw new Error('Missing token param');
    }

    //                ↓ how can I mock this fetch?
    const res = await fetch('https://api.example.com', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        token: params.token,
      }),
    });
  
    const json = await data.json();
  
    if (!json.session) {
      throw new Error('No session available');
    }

    // success!
    return NextResponse.redirect(new URL('/dashboard', request.url));
  } catch (error) {
    return NextResponse.redirect(new URL('/login', request.url));
  }
}
```

How can I mock the result of that fetch used inside my handler, so I can test different redirects?

```tsx
describe('setup link route handler', () => {
  it('redirects to 404 if token param is missing', async () => {
    await testApiHandler({
      appHandler,
      params: { token: '' },
      async test({ fetch }) {
        const res = await fetch({ method: 'GET' });
        expect(res.status).toBe(307);
        expect(res.headers.get('location')).toEndWith('/not-found');
      },
    });
  });

  it('redirects to 404 if token param is missing', async () => {
    await testApiHandler({
      appHandler,
      params: { token: '12345' },
      async test({ fetch }) {
        const res = await fetch({ method: 'GET' });
        expect(res.status).toBe(307);
        expect(res.headers.get('location')).toEndWith('/not-found');
      },
    });
  });
});
```

I tried using [msw](https://mswjs.io/) to setup some mock results, but that does not seem to do anything, which I believe is due to how Next mocks the fetch globally.

I would love some guidance as I've been knocking my head trying to test route handlers.

**Comment by Xunnamius** [comment, marked as answer]
Hey there! Unfortunately, it seems like you're running into [the incompatibility between MSW and the Next.js App Router](https://github.com/mswjs/msw/issues/1644). It looks thorny, but there are [some ways](https://github.com/Xunnamius/next-test-api-route-handler?tab=readme-ov-file#working-around-the-app-router-patching-the-global-fetch-function) you can restore the Native/MSW fetch if you don't want to use Next's.

**Reply by lucasmotta** [reply]
Thanks for the quick reply! I did tried that option too, but does not look like msw can intercept the fetch request still. The good news is that next 15 won't patch the global fetch anymore, which is music to my ears.

**Reply by Xunnamius** [reply]
Ah, this is great news. Thanks for the heads up on Next 15. Hopefully soon this won't be an issue we have to work around.
